services/api/service.go

```diff
@@ -1752,6 +1755,401 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
+			return
+		} else if errors.Is(err, datastore.ErrPastSlotAlreadyDelivered) {
+			// BAD VALIDATOR, 2x GETPAYLOAD FOR PAST SLOT
+			log.Warn("validator called getPayload for past slot")
+			api.RespondError(w, http.StatusBadRequest, "payload for this slot was already delivered")
+			return
+		} else if errors.Is(err, redis.TxFailedErr) {
+			// BAD VALIDATOR, 2x GETPAYLOAD + RACE
+			log.Warn("validator called getPayload twice (race)")
+			api.RespondError(w, http.StatusBadRequest, "payload for this slot was already delivered (race)")
+			return
+		}
+		log.WithError(err).Error("redis.CheckAndSetLastSlotAndHashDelivered failed")
+	}
+
+	// Handle early/late requests
+	if msIntoSlot < 0 {
+		// Wait until slot start (t=0) if still in the future
+		_msSinceSlotStart := time.Now().UTC().UnixMilli() - int64(slotStartTimestamp*1000) //nolint:gosec
+		if _msSinceSlotStart < 0 {
+			delayMillis := _msSinceSlotStart * -1
+			log = log.WithField("delayMillis", delayMillis)
+			log.Info("waiting until slot start t=0")
+			time.Sleep(time.Duration(delayMillis) * time.Millisecond)
+		}
+	} else if getPayloadRequestCutoffMs > 0 && msIntoSlot > int64(getPayloadRequestCutoffMs) {
+		// Reject requests after cutoff time
+		log.Warn("getPayload sent too late")
+		api.RespondError(w, http.StatusBadRequest, fmt.Sprintf("sent too late - %d ms into slot", msIntoSlot))
+
+		go func() {
+			err := api.db.InsertTooLateGetPayload(uint64(slot), proposerPubkey.String(), blockHash.String(), slotStartTimestamp, uint64(receivedAt.UnixMilli()), uint64(decodeTime.UnixMilli()), uint64(msIntoSlot)) //nolint:gosec
+			if err != nil {
+				log.WithError(err).Error("failed to insert payload too late into db")
+			}
+		}()
+		return
+	}
+
+	// Check that BlindedBlockContent fields (sent by the proposer) match our known BlockContents
+	err = EqBlindedBlockContentsToBlockContents(payload, getPayloadResp)
+	if err != nil {
+		log.WithError(err).Warn("ExecutionPayloadHeader not matching known ExecutionPayload")
+		api.RespondError(w, http.StatusBadRequest, "invalid execution payload header")
+		return
+	}
+
+	// Publish the signed beacon block via beacon-node
+	timeBeforePublish := time.Now().UTC().UnixMilli()
+	log = log.WithField("timestampBeforePublishing", timeBeforePublish)
+	signedBeaconBlock, err := common.SignedBlindedBeaconBlockToBeaconBlock(payload, getPayloadResp)
+	if err != nil {
+		log.WithError(err).Error("failed to convert signed blinded beacon block to beacon block")
+		api.RespondError(w, http.StatusInternalServerError, "failed to convert signed blinded beacon block to beacon block")
+		return
+	}
+	code, err := api.beaconClient.PublishBlock(signedBeaconBlock) // errors are logged inside
+	if err != nil || (code != http.StatusOK && code != http.StatusAccepted) {
+		log.WithError(err).WithField("code", code).Error("failed to publish block")
+		api.RespondError(w, http.StatusBadRequest, "failed to publish block")
+		return
+	}
```

Comment:
Block publishing needs to be async. The API should return 202 success once we acknowledge the block is valid and do the block publishing async.